### DISCUSSION TITLE
Continued Live Transcription of Audio in Python

### DISCUSSION THREAD

**Discussion by sadmoody** [deepgram/community#24] [General help] [answered]
I've been using Deepgram's Python SDK to transcribe audio. I want a constant stream of transcription to be put through from Python, but it appears that the SDK doesn't really prompt with responses until you await finish.

I'm using pyaudio to listen to my input device, sending chunks of data as they get generated. The callbacks aren't triggered until I stop sending data and send the empty websocket message.

Is there a way to constantly send data in Python without needing to signal a stop periodically in order to get transcripts?

Thanks!

**Comment by geekchick** [comment]
Hi @sadmoody ! Thank you for your question, and we appreciate you! I'm Tonya, the Python Developer Advocate at Deepgram. To make sure I understand your question, are you asking if there's a way to continually receive data in python without terminating or stopping the program to get the transcripts?

**Reply by sadmoody** [reply]
Yes and no. The program doesn't need to terminate - I can get it going again, it's more sending the empty websocket message is what triggers the responses.

It means that I need to figure out when a sentence ends on my side, which is something that I hoped Deepgram could do for me.

**Reply by SandraRodgers** [reply]
Are you passing the parameter `punctuate: true` to Deepgram, as in

```py
try:
  deepgramLive = await deepgram.transcription.live({ 'punctuate': True })
except Exception as e:
  print(f'Could not open socket: {e}')
  return
```

**Reply by sadmoody** [reply]
Yes! Will removing it make a difference?

**Reply by SandraRodgers** [reply]
No, I think you should have it based on what I'm understanding. You also might need `endpointing=true`. Here is some documentation that might be related to your issue.

https://developers.deepgram.com/documentation/features/endpointing/
https://developers.deepgram.com/documentation/guides/understand-endpointing-interim-results/

If you would give a snippet of your code, that would be helpful! I think when Tonya gets back online, she'll probably be able to help you more! Let me know if the `endpointing: true` helps.

Sandra

**Reply by sadmoody** [reply]
I'm pretty sure endpointing is set to true since I was playing with the "is_final" and "speech_final" params and they were changing.

Cheers Sandra, I'll check that when I get home from work and post a snippet.

**Reply by sadmoody** [reply]
Here's the snippet @SandraRodgers - if the `while` loop is changed to the commented out `for` loop, then the callbacks are called.

    async def main():

        deepgram = Deepgram(DEEPGRAM_API_KEY)
        
        try:
            deepgramLive = await deepgram.transcription.live({
                'punctuate': True,
                'interim_results': False,
                'language': 'en-NZ',
                'sample_rate': 44100,
                'encoding': 'linear16'
            })
        except Exception as e:
            print(f'Could not open socket: {e}')
            return
        
        deepgramLive.registerHandler(deepgramLive.event.CLOSE, lambda c: print(f'Connection closed with code {c}.'))
        deepgramLive.registerHandler(deepgramLive.event.TRANSCRIPT_RECEIVED, print)

        RATE = 44100
        CHUNK = 1024
        RECORD_SECONDS = 3

        py_audio = pyaudio.PyAudio()
        audio_stream = py_audio.open(
            rate=RATE,
            channels=1,
            format=pyaudio.paInt16,
            input=True,
            output=False,
            frames_per_buffer=1024,
            input_device_index=1
        )

        #for _ in range(int(RATE / CHUNK * RECORD_SECONDS)):
        while True: 
            data = audio_stream.read(CHUNK)
            deepgramLive.send(data)

        await deepgramLive.finish()

    asyncio.run(main())

**Reply by michaeljolley** [reply]
Heya @sadmoody! Just to clarify, you should be receiving transcripts before the `finish()` method is called. Be sure to enable `interim_results`. If you only want to consider a phrase when the API is done thinking about it, only process the received transcript if the `is_final` property is True.

The `finish()` method just sends the final "close" byte to the API and signals it to send a finalize any outstanding transcripts.

**Reply by scottstephenson** [reply]
Yep, definitely try enabling `interim_results=true` via the query parameter/flag!

**Reply by sadmoody** [reply]
That's exactly the behaviour I'd expect @MichaelJolley but it isn't what's happening. I'm not interested in interim results, I want the finalised results as the engine outputs them. I'd already tried enabling interim results and the behaviour is still the same - The results come back only once I call finish() - otherwise, it's just a constant listening/sending cycle. It's either the close byte doing something magic, or the SDK is holding back messages until the finish method is called. I haven't had the time to properly dig around in the SDK to figure out which one it is.

@scottstephenson I tried enabling interim results now just to see what happens and it's the same behaviour.

**Reply by geekchick** [reply]
Hi @sadmoody, I did notice there's one piece of code that's in our repo example that's missing from your code at first glance but trying to reproduce your issue locally on my machine:

```
if not data:
      break
```
         
    So in your while loop:
    
  ```
  while True:
                data = await audio.content.readany()
                deepgramLive.send(data)

                # If no data is sent from the live stream, then break out of the loop.
                if not data:
                    break

    # Indicate that we've finished sending data by sending the customary zero-byte message to the Deepgram streaming endpoint, and wait until we get back the final summary metadata object
    await deepgramLive.finish() 
```

**Reply by sadmoody** [reply]
@geekchick - I never stop sending data, I'm using pyaudio to interface with a USB microphone.

I've verified that:

1. I am capturing audio correctly - I can save to a wave file and listen back to it
2. Deepgram understands the audio that I send - everything works fine when I stop sending data

For my application I don't want to stop listening to and sending data from my input device. I expect Deepgram to send data back to me as it's ready.

If you're trying to replicate my issue, you may need to select a different `input_device_index` value depending on where your input device shows up from pyaudio's perspective. Also change the `RATE` and the `sample_rate` to reflect the one on the device. I shouldn't have hard-coded those 😬 but this is just a hacking exercise at this stage to figure out integration with Deepgram

**Reply by geekchick** [reply]
Hi @sadmoody thank you for the heads up. I'm using PyAudio and the mic from my computer. I had to use a different `input_device_index` for my mic and was able to reproduce something similar to your issue. I was not able to get a transcript back with `while True:` but am getting one with `for _ in range(int(RATE / CHUNK * RECORD_SECONDS))`. I'm looking into this further and will send a response soon. Please feel free to reply in the meantime with any other comments or new findings. I apologize for the inconvenience. This is a first!

**Reply by sadmoody** [reply]
Awesome work @geekchick 🎉

I'll get out of your hair in that case since you'll have more visibility over where the issue might be. My project is a little weird in its entirety, but I'd assume that this specific function wouldn't be too uncommon.

No need to apologise - glad you're on the case! Ping me if you come up with anything.

**Reply by geekchick** [reply]
Thank you for your understanding @sadmoody ! I'll continue working today on troubleshooting this issue. Will be in touch soon.

**Reply by geekchick** [reply]
Hi @sadmoody I didn't forget you! I'm back to troubleshooting this issue today.

**Reply by sadmoody** [reply]
Have you managed to make any progress @geekchick ?

I've almost run out of other things to do on my project!

**Reply by sadmoody** [reply]
I had a showerthought about this for a workaround which I'll try in the next couple of days. I'm suspecting that there's something in Deepgram's engine that doesn't like the raw signal being sent. If we process the wave input into a different format before sending it, that might work for now.

It adds a bit of overhead, but it gets me going again. This problem occurs when recording from a microphone but not when streaming the bbc audio for example. If we process the microphone audio into whatever format the bbc audio data is, that should work.

**Comment by geekchick** [comment, marked as answer]
Hi @sadmoody ! Thank you for your patience. I grabbed a code snippet from one of my colleagues here at Deepgram. Hitting they API directly, they were able to get live streaming working. I'm going to attached the python file `pyaudio-streaming` and the `requirements.txt` file below. I'm playing around with this code to figure out if I can adapt it to the Python SDK. Can you let me know if you were able to run the code sample?

[requirements.txt](https://github.com/deepgram/community/files/9707064/requirements.txt)
[pyaudio-streaming.txt](https://github.com/deepgram/community/files/9707068/pyaudio-streaming.txt)

**Reply by sadmoody** [reply]
Thanks so much - will check later tonight!

**Reply by sadmoody** [reply]
We're in business! Works great on my end and is blazingly fast! Will integrate this for now, and then move to the sdk when you get around to it. This isn't a blocker for me any more.

Thanks so much @geekchick! @MichaelJolley, make sure you give "big ups" to Tonya and this mystery colleague at the next standup. You need to use the phrase "big ups" in the sentence. Not accepting any regional alternatives thank you.

**Reply by geekchick** [reply]
Hi @sadmoody ! I'm so happy to hear that worked for you! :)
